### PR TITLE
[Refactor] Improve loops API 1/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds functionalities to the loop similar to torch nn.Module for checkpointing.

Next PR: #8362

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/loops/base.py [hunk 1]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
         self.trainer: Optional['pl.Trainer'] = None
         self.restarting = False
+        self._loops = OrderedDict()
+        self._progress = OrderedDict()
+        self._num_parents: int = 0
+
+    @property
+    def has_parent(self) -> Optional[bool]:
+        """Whether the number of loop parents is not null"""
+        return self._num_parents > 0
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply pytorch_lightning/loops/base.py:63]: also, how is the leaf definition below supposed to work if the graph is not a tree?

[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:63]: Should we enforce only 1 parent - multiple children structure ?

File: pytorch_lightning/loops/base.py [hunk 2]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+    @property
+    def has_children(self) -> bool:
+        """Whether this loop has any children"""
+        loops = self.__dict__.get('_loops')
+        return len(loops) > 0
+
+    @property
+    def is_leaf(self) -> bool:
+        """Whether this loop is a children and has no children itself."""
+        return not self.has_children and self.has_parent
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: Done !

File: pytorch_lightning/loops/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: Done !

**Review by justusschock — COMMENTED** [review]
TBH: I am not sure why we really need most of the stuff here. I left some comments, but IMO all the information about parents and children is not really required.

File: pytorch_lightning/loops/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (justusschock) [reply pytorch_lightning/loops/base.py]: Is the parent information relevant at all? Don't we usually start at the root of the tree/dag and then follow it's paths?

File: pytorch_lightning/loops/base.py [hunk 3]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
         self.iteration_count: int = 0
         self.trainer: Optional['pl.Trainer'] = None
         self.restarting = False
+        self._loops = OrderedDict()
+        self._progress = OrderedDict()
+        self._num_parents: int = 0
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: Removed !

File: pytorch_lightning/loops/base.py [hunk 4]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+    @property
+    def is_leaf(self) -> bool:
+        """Whether this loop is a children and has no children itself."""
+        return not self.has_children and self.has_parent
+
+    @property
+    def loop_progress(self) -> Dict[str, Any]:
+        """Return the progress for the current loop and children loop."""
+        progress = {}
+        for n, p in self.__dict__.get('_progress').items():
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:81]: removed !

File: pytorch_lightning/loops/base.py [hunk 5]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+            progress[n] = p
+
+        loops = self.__dict__.get('_loops')
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:84]: removed !

File: pytorch_lightning/loops/base.py [hunk 6]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+
+    @property
+    def has_parent(self) -> Optional[bool]:
+        """Whether the number of loop parents is not null"""
+        return self._num_parents > 0
+
+    @property
+    def has_children(self) -> bool:
+        """Whether this loop has any children"""
+        loops = self.__dict__.get('_loops')
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:69]: removed !

File: pytorch_lightning/loops/base.py [hunk 7]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+
+        loops = self.__dict__.get('_loops')
+
+        if loops is not None:
+            for name, loop in loops.items():
+                progress[name] = ProgressDict(**loop.loop_progress)
+        return ProgressDict(**progress)
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        if isinstance(value, pl.Trainer):
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:92]: removed !

File: pytorch_lightning/loops/base.py [hunk 8]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+            # when assigning a Trainer to a loop, it will assign to its children too.
+            object.__setattr__(self, name, value)
+            for loop in self._loops.values():
+                object.__setattr__(loop, name, value)
+        elif isinstance(value, Loop):
+            if getattr(self, "__children__loops__", None) is not None and name not in self.__children__loops__:
+                raise MisconfigurationException(
+                    f"The current loop accept only {self.__children__loops__} as children attribute names."
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: Removed !

File: pytorch_lightning/loops/base.py [hunk 9]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+            if not is_contained:
+                self._loops[name] = value
+                value._num_parents += 1
+        elif isinstance(value, BaseProgress):
+            self._progress[name] = value
+        else:
+            object.__setattr__(self, name, value)
+
+    def __getattr__(self, name) -> Any:
+        loops = self.__dict__.get('_loops')
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:116]: removed !

File: pytorch_lightning/loops/base.py [hunk 10]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+                        )
+            if not is_contained:
+                self._loops[name] = value
+                value._num_parents += 1
+        elif isinstance(value, BaseProgress):
+            self._progress[name] = value
+        else:
+            object.__setattr__(self, name, value)
+
+    def __getattr__(self, name) -> Any:
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py:115]: removed !

File: pytorch_lightning/loops/base.py [hunk 11]
```diff
@@ -47,6 +53,92 @@ def __init__(self) -> None:
...
+            object.__setattr__(self, name, value)
+            for loop in self._loops.values():
+                object.__setattr__(loop, name, value)
+        elif isinstance(value, Loop):
+            if getattr(self, "__children__loops__", None) is not None and name not in self.__children__loops__:
+                raise MisconfigurationException(
+                    f"The current loop accept only {self.__children__loops__} as children attribute names."
+                )
+            is_contained = False
+            for loop_name, loop in self._loops.items():
```
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: removed !

File: pytorch_lightning/loops/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tchaton) [reply pytorch_lightning/loops/base.py]: Removed !

File: pytorch_lightning/loops/base.py [hunk 12]
```diff
@@ -45,9 +47,36 @@ class Loop(ABC):
 
     def __init__(self) -> None:
         self.iteration_count: int = 0
-        self.trainer: Optional['pl.Trainer'] = None
+        self._trainer: Optional['pl.Trainer'] = None
         self.restarting = False
 
+    @property
+    def loop_progress(self) -> Dict[str, Any]:
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py:54]: this is for internal use, right? so we can make it protected and distinguish from the other progress attribute? Otherwise it will be confusing that there are two public properties for progress

File: pytorch_lightning/loops/base.py [hunk 13]
```diff
@@ -142,9 +163,51 @@ def on_run_end(self) -> Any:
...
+        self.restarting = True
+
+    def __load(self, state_dict, restart_progress, prefix=''):
+        self._load_from_state_dict(state_dict, prefix, restart_progress)
+        for k, v in self.__dict__.items():
+            if isinstance(v, Loop):
+                v.__load(state_dict.copy(), restart_progress, prefix + k + '.')
+
+    def load_state_dict(self, state_dict: Dict, restart_progress: bool = True):
+        self.__load(state_dict.copy(), restart_progress)
```
Comment (awaelchli) [inline pytorch_lightning/loops/base.py]: do we need load? we could also implement it directly here?

File: CHANGELOG.md [hunk 14]
```diff
@@ -339,6 +339,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Removed deprecated `optimizer` argument in `LightningModule.manual_backward()`; Toggling optimizers in manual optimization should be done using `LightningModule.{un}toggle_optimizer()` ([#8287](https://github.com/PyTorchLightning/pytorch-lightning/pull/8287))
 
 
+- Removed `Loop restore` function to give more control for loop restart ([#8334](https://github.com/PyTorchLightning/pytorch-lightning/pull/8334))
```
Comment (awaelchli) [inline CHANGELOG.md:342]: I think we don't need to mention this here, since these changes were never in a release.

File: pytorch_lightning/loops/base.py [hunk 15]
```diff
@@ -45,8 +47,35 @@ class Loop(ABC):
...
+
+    @trainer.setter
+    def trainer(self, trainer: 'pl.Trainer'):
+        """Connect the Trainer to itself and all its children loops"""
+        if not isinstance(trainer, pl.Trainer):
+            raise MisconfigurationException(
+                f"Loop {self.__class__.__name__} should be connected to a `Trainer`, found: {trainer}."
+            )
+        self._trainer = trainer
+        for v in self.__dict__.values():
```
Comment (justusschock) [inline pytorch_lightning/loops/base.py:65]: should we do the assignment recursively? I would actually prefer to do this more explicitly in the trainer (and also checking the loops properties maybe).

cc @awaelchli

File: pytorch_lightning/loops/base.py [hunk 16]
```diff
@@ -85,14 +110,13 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
         Returns:
             the output of :attr:`on_run_end` (often outputs collected from each step of the loop)
         """
+        if self.trainer is None:
+            raise MisconfigurationException(f"The {self.__class__.__name__} Loop hasn't been attached to any Trainer.")
```
Comment (justusschock) [inline pytorch_lightning/loops/base.py:103]: I think we should not enforce that if the loop does not need the trainer actually (e.g. if it is a custom one)

**Review by justusschock — APPROVED** [review]
Looks good to me overall. Just some comments

File: pytorch_lightning/loops/base.py (hunk 15 above)
Comment (awaelchli) [reply pytorch_lightning/loops/base.py:65]: I think we need such a mechanism  one way or the other and the trainer assignment can't go through the connect() method if connect() is the interface the user has to link loops together. It's more flexible for the user not to have to create a trainer instance first to connect custom loops. Plus it's safer imo to not allow a user to set different trainer references for loops.

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/loops/base.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:103]: Yes, let's remove this check.

File: pytorch_lightning/loops/base.py (hunk 15 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:65]: closing this one

File: pytorch_lightning/loops/base.py (hunk 16 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:103]: removed !

**Review by SeanNaren — APPROVED** [review]

File: tests/loops/test_loop_state_dict.py [hunk 17]
```diff
@@ -40,15 +40,209 @@ def test_loops_state_dict_structure():
...
+    # yapf: disable
     expected = {
         "fit_loop": {
-            'epoch_loop': {
-                'batch_loop': {},
-                'val_loop': {},
+            "epoch_loop": {
+                "batch_loop": {
+                    "state_dict": {},
+                    "progress": {
```
Comment (carmocca) [inline tests/loops/test_loop_state_dict.py:50]: The progress state dicts shoild be replaced with `ANY` as we already test them in

https://github.com/PyTorchLightning/pytorch-lightning/blob/370fa67004a0132733df91976490b7ed284bc406/tests/trainer/test_progress.py#L138

Otherwise the tests should be merged

File: pytorch_lightning/loops/base.py [hunk 18]
```diff
@@ -88,11 +102,7 @@ def run(self, *args: Any, **kwargs: Any) -> Optional[Any]:
         if self.skip:
             return self.on_skip()
 
-        if self.restarting:
```
Comment (carmocca) [inline pytorch_lightning/loops/base.py:91]: So why do we not want the `restarting` property anymore?

And why did you leave it in the `__init__`?

File: tests/loops/test_loops.py [hunk 19]
```diff
@@ -57,18 +77,145 @@ def load_state_dict(self, state_dict: Dict) -> None:
...
+    loop_parent = Simple(1)
+    loop_child = Simple(2)
+    loop_parent.loop_child = loop_child
+    loop_parent.load_state_dict(state_dict)
+    assert loop_parent.progress.increment == 1
+    assert loop_parent.loop_child.progress.increment == 0
+
+    del loop_parent.loop_child
+    state_dict = loop_parent.state_dict()
+    assert state_dict == OrderedDict([('state_dict', {'a': 1}), ('progress', {'increment': 1})])
```
Comment (carmocca) [inline tests/loops/test_loops.py:212]: The state dicts of the Loop class do not return `OrderedDict`s. Should they?

If not, why are we testing them here?

File: tests/loops/test_loop_state_dict.py (hunk 17 above)
Comment (tchaton) [reply tests/loops/test_loop_state_dict.py:50]: I prefer to explicit show the structure of the `state dict`. This is easier for people to understand its true structure.

File: pytorch_lightning/loops/base.py (hunk 18 above)
Comment (tchaton) [reply pytorch_lightning/loops/base.py:91]: This is still used in reset `self.restarting`. Check https://github.com/PyTorchLightning/pytorch-lightning/pull/8364.

Adrian convinced me it was cleaner to make it manual.

self.restarting is automatically set to True by the Trainer when calling loop `load_state_dict` function.

File: tests/loops/test_loops.py (hunk 19 above)
Comment (tchaton) [reply tests/loops/test_loops.py:212]: Good question. Not anymore. This test should be updated.

File: pytorch_lightning/loops/base.py (hunk 18 above)
Comment (carmocca) [reply pytorch_lightning/loops/base.py:91]: Where exactly is that PR setting `restarting` on the loops?

Comment (tchaton) [reply pytorch_lightning/loops/base.py:91]: Hey @carmocca, right there: https://github.com/PyTorchLightning/pytorch-lightning/pull/8334/files#diff-4076755145b42e55247c8c12a7a7961a91f4d28a3e962c80a98161163bfa3ec9R205.

Each loop is going to load its own checkpoint and be put in restarting mode.
